# 4차 과제 (준비중)

In [1]:
import numpy as np
import pandas as pd

In [2]:
np.set_printoptions(precision=6, suppress=True)
PREVIOUS_MAX_ROWS = pd.options.display.max_rows # 원래 60이 기본.
pd.set_option("max_rows", 20)

In [3]:
from sklearn.datasets import load_boston
X = load_boston()

`X`의 자료형을 확인하라. 

In [4]:
type(X)

sklearn.utils.Bunch

`Bunch` 자료형을 `dict()` 자료형과 비교하여 설명하라. 

답)

`X`에 사용된 키(key)를 확인하라.

In [5]:
X.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

data와 target이 구분되어 있다. 

In [6]:
X.data

array([[  0.00632,  18.     ,   2.31   , ...,  15.3    , 396.9    ,
          4.98   ],
       [  0.02731,   0.     ,   7.07   , ...,  17.8    , 396.9    ,
          9.14   ],
       [  0.02729,   0.     ,   7.07   , ...,  17.8    , 392.83   ,
          4.03   ],
       ...,
       [  0.06076,   0.     ,  11.93   , ...,  21.     , 396.9    ,
          5.64   ],
       [  0.10959,   0.     ,  11.93   , ...,  21.     , 393.45   ,
          6.48   ],
       [  0.04741,   0.     ,  11.93   , ...,  21.     , 396.9    ,
          7.88   ]])

In [7]:
X.data.shape

(506, 13)

In [8]:
X.target[:50]

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4])

In [9]:
X.target.shape

(506,)

`X.data`는 미국 보스턴 시의 구역별 통계자료를 갖고 있으며, (506, 13) 모양의 2차원 어레이다.
각각의 열(column)은 아래 특성을 가리킨다. 

| 특성 | 의미 |
|:------|:---------|
| CRIM  | 구역별 1인당 범죄율 |
| ZN    | 25,000 평방 피트 이상의 주거 구역 비율 |
| INDUS | 구역별 비 소매 사업 에이커(acre) 비율 |
| CHAS  | Charles River 더미 변수(구역이 강 경계에 닿으면 1, 아니면 0) |
| NOX   | 산화 질소 농도(1000만분 율) |
| RM    | 주택 당 평균 방 수 |
| AGE   | 소유주가 살고 있는 1940년 이전에 지어진 건물 비율 |
| DIS   | 보스턴 고용 센터 다섯 곳 까지의 가중 거리 |
| RAD   | 방사형 고속도로 접근성 지수 |
| TAX   | 1만달러당 전체 가지 재산 세율 |
| PTRATIO | 구역별 학생-교사 비율 |
| B     | 1000(Bk - 0.63)^2 (Bk 구역별 흑인 비율) |
| LSTAT | 구역별 낮은 지위 인구 비율 |

`X.target`는 미국 보스턴 시의 506개 구역별 소유주 거주 주택의 중간 가격(단위: 1,000달러)이다. 

| 특성 | 의미 |
|:------|:---------|
| MEDV  | 소유주 거주 주택의 중간 가격(단위: 1,000달러) |

이제 `X.data`를 `boston` 변수에 할당한다.

In [10]:
boston = X.data
boston

array([[  0.00632,  18.     ,   2.31   , ...,  15.3    , 396.9    ,
          4.98   ],
       [  0.02731,   0.     ,   7.07   , ...,  17.8    , 396.9    ,
          9.14   ],
       [  0.02729,   0.     ,   7.07   , ...,  17.8    , 392.83   ,
          4.03   ],
       ...,
       [  0.06076,   0.     ,  11.93   , ...,  21.     , 396.9    ,
          5.64   ],
       [  0.10959,   0.     ,  11.93   , ...,  21.     , 393.45   ,
          6.48   ],
       [  0.04741,   0.     ,  11.93   , ...,  21.     , 396.9    ,
          7.88   ]])

열(column)별 이름, 즉 특성별 이름을 지정하면서 판다스의 데이터프레임으로 형변환한다.
특성별 이름은 `X`의 `feature_names` 속성에 저장되어 있다.

In [11]:
X.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [12]:
boston = pd.DataFrame(X.data, columns=X.feature_names)
boston

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

특성별 간단한 정보는 다음과 같다.

In [13]:
boston.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


수치형 데이터의 통계 자료는 다음과 같다.

In [14]:
boston.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  506.000000  
mean    12.653063  
std      7.141062  
min      1.730000  
25%      6.950000  
50%     11.360000  
75%     16.955000  
max     37.970000

원래 결측치가 없지만 아래처럼 고의로 5개의 결측치를 입력한다.
대상은 주택당 평균 방수 특성(RM)의 100번에서 104번행이다.

In [15]:
boston['RM'][100:105] = np.nan

In [16]:
boston.loc[97:107]

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
97   0.12083  0.0   2.89   0.0  0.445  8.069  76.0  3.4952  2.0  276.0   
98   0.08187  0.0   2.89   0.0  0.445  7.820  36.9  3.4952  2.0  276.0   
99   0.06860  0.0   2.89   0.0  0.445  7.416  62.5  3.4952  2.0  276.0   
100  0.14866  0.0   8.56   0.0  0.520    NaN  79.9  2.7778  5.0  384.0   
101  0.11432  0.0   8.56   0.0  0.520    NaN  71.3  2.8561  5.0  384.0   
102  0.22876  0.0   8.56   0.0  0.520    NaN  85.4  2.7147  5.0  384.0   
103  0.21161  0.0   8.56   0.0  0.520    NaN  87.4  2.7147  5.0  384.0   
104  0.13960  0.0   8.56   0.0  0.520    NaN  90.0  2.4210  5.0  384.0   
105  0.13262  0.0   8.56   0.0  0.520  5.851  96.7  2.1069  5.0  384.0   
106  0.17120  0.0   8.56   0.0  0.520  5.836  91.9  2.2110  5.0  384.0   
107  0.13117  0.0   8.56   0.0  0.520  6.127  85.2  2.1224  5.0  384.0   

     PTRATIO       B  LSTAT  
97      18.0  396.90   4.21  
98      18.0  393.53   3.57  
99      18.0  396.90   6.19  
100     20.9  394.76   9.42  
101     20.9  395.58   7.67  
102     20.9   70.80  10.63  
103     20.9  394.47  13.44  
104     20.9  392.69  12.33  
105     20.9  394.05  16.47  
106     20.9  395.67  18.66  
107     20.9  387.69  14.09

그러면 info()의 결과가 다음처럼 변한다.

In [17]:
boston.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       501 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


다섯 개의 값만 결측치임을 확인하라. 

In [18]:
boston.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         5
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

결측치가 들어있는 행만 확인하면 다음과 같다.

In [19]:
boston_with_nan = boston[boston.RM.isnull()].copy()
boston_with_nan

CRIM   ZN  INDUS  CHAS   NOX  RM   AGE     DIS  RAD    TAX  PTRATIO  \
100  0.14866  0.0   8.56   0.0  0.52 NaN  79.9  2.7778  5.0  384.0     20.9   
101  0.11432  0.0   8.56   0.0  0.52 NaN  71.3  2.8561  5.0  384.0     20.9   
102  0.22876  0.0   8.56   0.0  0.52 NaN  85.4  2.7147  5.0  384.0     20.9   
103  0.21161  0.0   8.56   0.0  0.52 NaN  87.4  2.7147  5.0  384.0     20.9   
104  0.13960  0.0   8.56   0.0  0.52 NaN  90.0  2.4210  5.0  384.0     20.9   

          B  LSTAT  
100  394.76   9.42  
101  395.58   7.67  
102   70.80  10.63  
103  394.47  13.44  
104  392.69  12.33

In [20]:
boston_without_nan = boston.dropna(how='any').copy()
boston_without_nan.shape

(501, 13)

평균값으로 대체할 때의 결과는?

In [21]:
boston_with_nan.fillna(boston.mean())

CRIM   ZN  INDUS  CHAS   NOX       RM   AGE     DIS  RAD    TAX  \
100  0.14866  0.0   8.56   0.0  0.52  6.28305  79.9  2.7778  5.0  384.0   
101  0.11432  0.0   8.56   0.0  0.52  6.28305  71.3  2.8561  5.0  384.0   
102  0.22876  0.0   8.56   0.0  0.52  6.28305  85.4  2.7147  5.0  384.0   
103  0.21161  0.0   8.56   0.0  0.52  6.28305  87.4  2.7147  5.0  384.0   
104  0.13960  0.0   8.56   0.0  0.52  6.28305  90.0  2.4210  5.0  384.0   

     PTRATIO       B  LSTAT  
100     20.9  394.76   9.42  
101     20.9  395.58   7.67  
102     20.9   70.80  10.63  
103     20.9  394.47  13.44  
104     20.9  392.69  12.33

boston_with_nan의 값은 그대로 이다.

In [22]:
boston_with_nan

CRIM   ZN  INDUS  CHAS   NOX  RM   AGE     DIS  RAD    TAX  PTRATIO  \
100  0.14866  0.0   8.56   0.0  0.52 NaN  79.9  2.7778  5.0  384.0     20.9   
101  0.11432  0.0   8.56   0.0  0.52 NaN  71.3  2.8561  5.0  384.0     20.9   
102  0.22876  0.0   8.56   0.0  0.52 NaN  85.4  2.7147  5.0  384.0     20.9   
103  0.21161  0.0   8.56   0.0  0.52 NaN  87.4  2.7147  5.0  384.0     20.9   
104  0.13960  0.0   8.56   0.0  0.52 NaN  90.0  2.4210  5.0  384.0     20.9   

          B  LSTAT  
100  394.76   9.42  
101  395.58   7.67  
102   70.80  10.63  
103  394.47  13.44  
104  392.69  12.33

어떻게 하면 nan 을 boston_with_nan에서 모두 대체하는가? 

In [23]:
boston_with_nan.fillna(boston.mean(), inplace=True)

In [24]:
boston_with_nan

CRIM   ZN  INDUS  CHAS   NOX       RM   AGE     DIS  RAD    TAX  \
100  0.14866  0.0   8.56   0.0  0.52  6.28305  79.9  2.7778  5.0  384.0   
101  0.11432  0.0   8.56   0.0  0.52  6.28305  71.3  2.8561  5.0  384.0   
102  0.22876  0.0   8.56   0.0  0.52  6.28305  85.4  2.7147  5.0  384.0   
103  0.21161  0.0   8.56   0.0  0.52  6.28305  87.4  2.7147  5.0  384.0   
104  0.13960  0.0   8.56   0.0  0.52  6.28305  90.0  2.4210  5.0  384.0   

     PTRATIO       B  LSTAT  
100     20.9  394.76   9.42  
101     20.9  395.58   7.67  
102     20.9   70.80  10.63  
103     20.9  394.47  13.44  
104     20.9  392.69  12.33

In [25]:
boston1 = boston.copy()

In [26]:
boston1.fillna(boston1.mean(), inplace=True)

In [27]:
boston1.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

AGE 특성의 값들을 만 나이로 고쳐라. 힌트: np.floor() 함수 활용

In [28]:
boston1['AGE_floor'] = boston1.AGE.apply(np.floor).apply(int)

AGE를 삭제하라.

In [29]:
boston1.drop('AGE', axis=1)

CRIM    ZN  INDUS  CHAS    NOX     RM     DIS  RAD    TAX  PTRATIO  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  4.0900  1.0  296.0     15.3   
1    0.02731   0.0   7.07   0.0  0.469  6.421  4.9671  2.0  242.0     17.8   
2    0.02729   0.0   7.07   0.0  0.469  7.185  4.9671  2.0  242.0     17.8   
3    0.03237   0.0   2.18   0.0  0.458  6.998  6.0622  3.0  222.0     18.7   
4    0.06905   0.0   2.18   0.0  0.458  7.147  6.0622  3.0  222.0     18.7   
..       ...   ...    ...   ...    ...    ...     ...  ...    ...      ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  2.4786  1.0  273.0     21.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  2.2875  1.0  273.0     21.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  2.1675  1.0  273.0     21.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  2.3889  1.0  273.0     21.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  2.5050  1.0  273.0     21.0   

          B  LSTAT  AGE_floor  
0    396.90   4.98         65  
1    396.90   9.14         78  
2    392.83   4.03         61  
3    394.63   2.94         45  
4    396.90   5.33         54  
..      ...    ...        ...  
501  391.99   9.67         69  
502  396.90   9.08         76  
503  396.90   5.64         91  
504  393.45   6.48         89  
505  396.90   7.88         80  

[506 rows x 13 columns]

In [30]:
boston1['AGE'] = boston1.AGE_floor

In [31]:
boston1.head()

CRIM    ZN  INDUS  CHAS    NOX     RM  AGE     DIS  RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575   65  4.0900  1.0  296.0     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421   78  4.9671  2.0  242.0     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185   61  4.9671  2.0  242.0     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998   45  6.0622  3.0  222.0     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147   54  6.0622  3.0  222.0     18.7   

        B  LSTAT  AGE_floor  
0  396.90   4.98         65  
1  396.90   9.14         78  
2  392.83   4.03         61  
3  394.63   2.94         45  
4  396.90   5.33         54